### DISCUSSION TITLE
GetStaticProps redirection not possible during rerender

### DISCUSSION THREAD

**Discussion by trm217** [vercel/next.js#42677] [Ideas]
Returning a redirect from within a getStaticProps has been possible for quite a while, but today I noticed a issue, when trying to always return a redirect from getStaticProps.

## My use-case

Let me explain my use-case first. We publish articles everyday with a new url and we want a static url that then redirects to the most recent article. Now the most recent article-url has to be fetched from our server using GraphQL. 

This URL only chances at most once per day, so I could perfectly cache the result of the first server-request and reuse it for all users.
I then added a page (the component returns null, since I always want to redirect) with a getStaticProps function, that fetches the redirect-target from the api and then returns the redirection (with a revalidation time relative to the start of the next day).
While trying to build our website with this setup, I got the following error:

```
Error occurred prerendering page "/generic". Read more: https://nextjs.org/docs/messages/prerender-error
@orbiting/www-app:build: Error: `redirect` can not be returned from getStaticProps during prerendering (/generic)
```

So it seems I can't only return a redirect within a getStaticProps-function, I must at least return actual props during at build-time. I suppose this happens because Next.js expects a initial page to be given during rendering, however I feel like it would be great if we could also leverage the performance benefits of getStaticProps when it comes to GetStaticProps

## Alternative solutions

One solution would be to implement this in the middleware, however this would have a rather large overhead since it would have to fetch the redirect-target from the API with ever single user-request. 
Same goes when replacing getStaticProps with getServerSideProps.

While both of these are an option, I would rather have it be cached for all users.

**Comment by icyJoseph** [comment]
Yeah, build time redirect from `getStaticProps` is not possible.

However, we could try to limit the pages touched by `getStaticProps`, so that we don't need to run into the redirect situation.

You could use `getStaticPaths` I guess, to collect all valid URLs, and return them. You could even be bolder and return and empty list of paths, https://nextjs.org/docs/basic-features/data-fetching/get-static-paths#generating-paths-on-demand, with fallback set to true, with proper handling of the fallback page, https://nextjs.org/docs/api-reference/data-fetching/get-static-paths#fallback-pages.

What do you think?

**Reply by trm217** [reply]
Let me explain my situation: We are routing article for an online magazin with the path `/[...path]`. The CMS can export paths of various sizes from `/cms-page` to `/[yyyy]/[mm]/[dd]/[web-slug]`.
(For `[...path]`, we use getStaticPaths, however since we have thousands of pages, we don't generate them at build time, but instead as soon as a user requests a specific url)

So whenever we need custom logic for a page, we manually add that page to the pages folder to prevent the CMS from ever overriding it. (Of course it would be possible to define blocked routes in the CMS, but I heavily dislike that approach)

At the moment we reverted to using the middleware to do the redirection, but as said this is suboptimal, since the looked up redirection target is the same for ever user. At peek time this leads to thousands requests that yield the exact same redirect.
Of course we could add the logic to the getStaticProps inside `/[...path]`, however this will make the already extremely complex getStaticProps function in there even less comprehendible.

I would really encourage thinking about allowing returning redirects in general at all time from within getStaticProps. This would make SSG even more versatile and allow for performant dynamic rewrites.

**Reply by icyJoseph** [reply]
Am I missing something here? Your project doesn't generate the pages at build time, then why is it having problems with redirect from GetStaticPaths?

**Reply by trm217** [reply]
As said, I'd rather have a separate page handling this case, since it also guarantees, that potential path overrides due to the CMS won't occur.
And anyway, I feel that being able to return a redirect in SSG at runtime, but not at build time seems very odd to me. Another solution would be just allow for the generation of the page to be delayed until after build-time.
